iOS Coding/Network/NetworkManager.swift

```diff
@@ -0,0 +1,79 @@
+    }
+    
+    func executeRequest(requestConfiguration: RequestConfigurationProtocol, responseHandler: ResponseHandlerProtocol = ResponseHandler(), dataRequestCompletion: DataRequestCompletion?) {
+        executeRequest(requestConfiguration: requestConfiguration, responseHandler: responseHandler, responseModel: NilDecodable.self, modelRequestCompletion: nil, dataRequestCompletion: dataRequestCompletion)
+    }
+    
+    private func executeRequest<T: Decodable>(requestConfiguration: RequestConfigurationProtocol, responseHandler: ResponseHandlerProtocol = ResponseHandler(), responseModel: T.Type, modelRequestCompletion: ModelRequestCompletion?, dataRequestCompletion: DataRequestCompletion?) {
+        requestConfiguration.generateRequest { result in
+            withExtendedLifetime(requestConfiguration) {
+                switch result {
+                case .success(let request):
+                    let task = self.dataTask(request: request, responseHandler: responseHandler, responseModel: responseModel, modelRequestCompletion: modelRequestCompletion, dataRequestCompletion: dataRequestCompletion)
+                    task.resume()
+                case .failure(let error):
+                    modelRequestCompletion?(.failure(error))
+                    dataRequestCompletion?(.failure(error))
+                }
+            }
+        }
+    }
+    
+    func dataTask<T: Decodable>(request: URLRequest, responseHandler: ResponseHandlerProtocol, responseModel: T.Type, modelRequestCompletion: ModelRequestCompletion?, dataRequestCompletion: DataRequestCompletion?) -> URLSessionDataTask {
+        let task = session.dataTask(with: request) { data, response, error in
+            if let error = error {
+                modelRequestCompletion?(.failure(NetworkError.networkFailure(error: error)))
+                dataRequestCompletion?(.failure(NetworkError.networkFailure(error: error)))
+                return
+            }
+            
+            do {
+                try responseHandler.validateResponse(response: response, data: data)
+            } catch (let validateResponseError) {
+                modelRequestCompletion?(.failure(validateResponseError))
+                dataRequestCompletion?(.failure(validateResponseError))
+                return
+            }
+            
+            if !(responseModel is NilDecodable.Type), let data = data {
+                do {
+                    let dataObject: Decodable = try JSONDecoder().decode(responseModel, from: data)
+                    let dataResponse = DataResponse(data: dataObject, urlResponse: response)
+                    modelRequestCompletion?(.success(dataResponse))
+                    return
+                } catch {
+                    modelRequestCompletion?(.failure(NetworkError.decodeModelFailure(error: error)))
```

Comment:
Missing a return here. If an error occurs then the modelRequestCompletion will get called twice, with the last one being a success.